README.md
# Scala JSON Schema
[![Build](https://github.com/andyglow/scala-jsonschema/actions/workflows/ci.yml/badge.svg?event=push)](https://github.com/andyglow/scala-jsonschema/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/andyglow/scala-jsonschema/branch/master/graph/badge.svg?token=e8E62N8PiJ)](https://codecov.io/gh/andyglow/scala-jsonschema)
![](https://img.shields.io/static/v1.svg?label=Supported&nbsp;Scala&message=2.11,2.12,2.13&color=gray)
[![Version](https://img.shields.io/badge/dynamic/json.svg?label=Released&nbsp;Version&query=%24.response.docs%5B0%5D.latestVersion&url=https%3A%2F%2Fsearch.maven.org%2Fsolrsearch%2Fselect%3Fq%3Dscala-jsonschema-core_2.13%26start%3D0%26rows%3D1)](https://search.maven.org/artifact/com.github.andyglow/scala-jsonschema-core_2.13/)

*SBT dependencies:*

Main module:

```scala
libraryDependencies += "com.github.andyglow" %% "scala-jsonschema" % <version> // <-- required
```

Other libraries:

```scala
libraryDependencies ++= Seq(
  "com.github.andyglow" %% "scala-jsonschema-core" % <version>,              // <-- transitive
  "com.github.andyglow" %% "scala-jsonschema-macros" % <version> % Provided, // <-- transitive
  // json bridge. pick one
  "com.github.andyglow" %% "scala-jsonschema-play-json" % <version>,         // <-- optional
  "com.github.andyglow" %% "scala-jsonschema-spray-json" % <version>,        // <-- optional
  "com.github.andyglow" %% "scala-jsonschema-circe-json" % <version>,        // <-- optional
  "com.github.andyglow" %% "scala-jsonschema-json4s-json" % <version>,       // <-- optional
  "com.github.andyglow" %% "scala-jsonschema-ujson" % <version>,             // <-- optional
  // joda-time support
  "com.github.andyglow" %% "scala-jsonschema-joda-time" % <version>,         // <-- optional
  // cats support
  "com.github.andyglow" %% "scala-jsonschema-cats" % <version>,              // <-- optional
  // refined support
  "com.github.andyglow" %% "scala-jsonschema-refined" % <version>,           // <-- optional
  // enumeratum support
  "com.github.andyglow" %% "scala-jsonschema-enumeratum" % <version>,        // <-- optional
  // zero-dependency json and jsonschema parser
  "com.github.andyglow" %% "scala-jsonschema-parser" % <version>             // <-- optional
)
```


## Generate JSON Schema from Scala classes

The goal of this library is to make JSON Schema generation done the way all popular JSON reading/writing libraries do.
Inspired by Coursera Autoschema but uses `Scala Macros` instead of `Java Reflection`.

## Features
- Supports Json Schema `draft-04`, `draft-06`, `draft-07`, `draft-09`, `draft-12`
- Supports `case classes`
- Supports `value classes`
- Supports `sealed trait enums`
- Supports `sealed trait case classes`
- Supports `recursive types`
- Supports `scala.Enumeration`
- Treats `scala.Option` as optional fields
- As well as treats fields with `default values` as optional
- Any `Iterable` is treated as `array`
- Pluggable [Joda-Time](https://github.com/JodaOrg/joda-time) Support
- Pluggable [Cats](https://github.com/typelevel/cats) Support
- Pluggable [Refined](https://github.com/fthomas/refined) Support
- Pluggable [Enumeratum](https://github.com/lloydmeta/enumeratum) Support
- Supports generic data types

### Types supported out of the box
- `Boolean`
- Numeric
    - `Short`
    - `Int`
    - `Char`
    - `Double`
    - `Float`
    - `Long`
    - `BigInt`
    - `BigDecimal`
- `String`
- Date Time
    - `java.util.Date`
    - `java.sql.Timestamp`
    - `java.time.Instant`
    - `java.time.LocalDateTime`
    - `java.sql.Date`
    - `java.time.LocalDate`
    - `java.sql.Time`
    - `java.time.LocalTime`
- with JodaTime module imported
    - `org.joda.time.Instant`
    - `org.joda.time.DateTime`
    - `org.joda.time.LocalDateTime`
    - `org.joda.time.LocalDate`
    - `org.joda.time.LocalTime`
- with Cats module imported
    - `cats.data.NonEmptyList`
    - `cats.data.NonEmptyVector`
    - `cats.data.NonEmptySet`
    - `cats.data.NonEmptyChain`
    - `cats.data.NonEmptyMap`
    - `cats.data.NonEmptyStream` (for scala 2.11, 2.12)
    - `cats.data.NonEmptyLazyList` (for scala 2.13)
    - `cats.data.OneAnd`
- with Refined module imported you can refine original types with these
    - boolean
        - `eu.timepit.refined.boolean.And`
        - `eu.timepit.refined.boolean.Or`
        - `eu.timepit.refined.boolean.Not`
    - string
        - `eu.timepit.refined.collection.Size`
        - `eu.timepit.refined.collection.MinSize`
        - `eu.timepit.refined.collection.MaxSize`
        - `eu.timepit.refined.collection.Empty`
        - `eu.timepit.refined.collection.NonEmpty`
        - `eu.timepit.refined.string.Uuid`
        - `eu.timepit.refined.string.Uri`
        - `eu.timepit.refined.string.Url`
        - `eu.timepit.refined.string.IPv4`
        - `eu.timepit.refined.string.IPv6`
        - `eu.timepit.refined.string.Xml`
        - `eu.timepit.refined.string.StartsWith`
        - `eu.timepit.refined.string.EndsWith`
        - `eu.timepit.refined.string.MatchesRegex`
        - `eu.timepit.refined.string.Trimmed`
    - number
        - `eu.timepit.refined.numeric.Positive`
        - `eu.timepit.refined.numeric.Negative`
        - `eu.timepit.refined.numeric.NonPositive`
        - `eu.timepit.refined.numeric.NonNegative`
        - `eu.timepit.refined.numeric.Greather`
        - `eu.timepit.refined.numeric.Less`
        - `eu.timepit.refined.numeric.GreaterEqual`
        - `eu.timepit.refined.numeric.LessEqual`
        - `eu.timepit.refined.numeric.Divisable`
    - collection
        - `eu.timepit.refined.collection.Size`
        - `eu.timepit.refined.collection.MinSize`
        - `eu.timepit.refined.collection.MaxSize`
        - `eu.timepit.refined.collection.Empty`
        - `eu.timepit.refined.collection.NonEmpty`
- with Enumeratum module enabled
    - `enums` based on `EnumEntry`/`Enum`            
    - `enums` based on `ValueEnumEntry`/`ValueEnum`            
- Misc
    - `java.util.UUID`
    - `java.net.URL`
    - `java.net.URI`
- Collections
    - String Map (eg. `Map[String, T]`)
    - Int Map (eg. `Map[Int, T]`)
    - `Iterable[T]`
- Sealed Trait hierarchy of case objects (Enums)
- Case Classes
    - default value
- Sealed Trait hierarchy of case classes
- Value Classes    

## Example
Suppose you have defined this data structures
```scala
sealed trait Gender

object Gender {

    case object Male extends Gender

    case object Female extends Gender
}

case class Company(name: String)

case class Car(name: String, manufacturer: Company)

case class Person(
    firstName: String,
    middleName: Option[String],
    lastName: String,
    gender: Gender,
    birthDay: java.time.LocalDateTime,
    company: Company,
    cars: Seq[Car])
```

Now you have several ways to specify your schema.

### In-Lined
In simple words in-lined mode means you will have no `definitions`. Type you want to use as source for schema will
be represented in json schema without reusable data blocks.  
```scala
import json._

val personSchema: json.Schema[Person] = Json.schema[Person]
```
As result you will receive this:
```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "middleName": {
      "type": "string"
    },
    "cars": {
      "type": "array",
      "items": {
        "type": "object",
        "additionalProperties": false,
        "properties": {
          "name": {
            "type": "string"
          },
          "manufacturer": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
              "name": {
                "type": "string"
              }
            },
            "required": [
              "name"
            ]
          }
        },
        "required": [
          "name",
          "manufacturer"
        ]
      }
    },
    "company": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "name": {
          "type": "string"
        }
      },
      "required": [
        "name"
      ]
    },
    "lastName": {
      "type": "string"
    },
    "firstName": {
      "type": "string"
    },
    "birthDay": {
      "type": "string",
      "format": "date-time"
    },
    "gender": {
      "type": "string",
      "enum": [
        "Male",
        "Female"
      ]
    }
  },
  "required": [
    "company",
    "lastName",
    "birthDay",
    "gender",
    "firstName",
    "cars"
  ]
}
```

### Regular
Schema generated in Regular mode will contain so many `definitions` so many separated definitions you provide.
Lets take a look at example code: 

```scala
import json._

implicit val genderSchema: json.Schema[Gender] = Json.schema[Gender]

implicit val companySchema: json.Schema[Company] = Json.schema[Company]

implicit val carSchema: json.Schema[Car] = Json.schema[Car]

implicit val personSchema: json.Schema[Person] = Json.schema[Person]
```
Here we defined, besides Person schema, gender, company and car schemas. 
The result will be looking this way then.
```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "middleName": {
      "type": "string"
    },
    "cars": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/com.github.andyglow.jsonschema.ExampleMsg.Car"
      }
    },
    "company": {
      "$ref": "#/definitions/com.github.andyglow.jsonschema.ExampleMsg.Company"
    },
    "lastName": {
      "type": "string"
    },
    "firstName": {
      "type": "string"
    },
    "birthDay": {
      "type": "string",
      "format": "date-time"
    },
    "gender": {
      "$ref": "#/definitions/com.github.andyglow.jsonschema.ExampleMsg.Gender"
    }
  },
  "required": [
    "company",
    "lastName",
    "birthDay",
    "gender",
    "firstName",
    "cars"
  ],
  "definitions": {
    "com.github.andyglow.jsonschema.ExampleMsg.Company": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "name": {
          "type": "string"
        }
      },
      "required": [
        "name"
      ]
    },
    "com.github.andyglow.jsonschema.ExampleMsg.Car": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "name": {
          "type": "string"
        },
        "manufacturer": {
          "$ref": "#/definitions/com.github.andyglow.jsonschema.ExampleMsg.Company"
        }
      },
      "required": [
        "name",
        "manufacturer"
      ]
    },
    "com.github.andyglow.jsonschema.ExampleMsg.Gender": {
      "type": "string",
      "enum": [
        "Male",
        "Female"
      ]
    }
  }
}
```

## Definitions / References
There is a couple of ways to specify reference of schema.
1. It could be generated from type name (including type args)
2. You can do it yourself. It is useful when you want to provide couple of schemas with same type but with different validation rules.

So originally you use
```scala
import json._

implicit val someStrSchema: json.Schema[String] = Json.schema[String]


CHANGELOG.md
- **0.7.10**
  - draft 2020-12 implemented
  - dependencies updated 
- **0.7.1**
  - draft 2019-09 implemented
  - initial json-diff functionality added  
  ...
- **0.6.3**
  - fixed scala.None handling for the default values of case class parameters (scala 2.11 is not supported this kind of functionality)
- **0.6.2**
  - added annotations for
    - definition enforcement
    - discriminator/discriminatorKey  
- **0.6.1**
  - added support for scala.Enumeration
  - default values won't be inferred for scala.None
- **0.6.0**
  ...
- **0.2.7-M1**
  - Add initial code for supporting #18. Use scaladoc for descriptions. 
    Added support for case classes and fields.
  - todo:
    - add scaladoc support for sum types
    - here is some problem. comments live in source files and get erased after compilation, so resulted bytecode 
      doesn't have this info and thus schemas derived from such models won't have descriptions.
      this might probably be solved by packing scaladocs into annotations by macro or compiler plugin and
      have those reused my SchemaMacro. Here is an example of something like this: 
      https://medium.com/@takezoe/a-compiler-plugin-makes-scaladoc-readable-at-runtime-aecbebccb794 
- **0.2.6**
  - hot-fix release. disable mistakingly left enabled debug mode for Refined macro
- **0.2.5**
  - Improve Schema matching. 
    Since now Ref and Validations are taken into an account during schema comparision.
  - Cats version upgraded to 2.1.1 for scala 2.12 and greater  
  - Support Refined types (0.9.13 for scala 2.12, 2.13, 0.9.12 for scala 2.11) 
- **0.2.4**
  - Support Cats data types (non-empty-*)
  - Support generic product types
  - Module `scala-jsonschema-api` renamed into `scala-jsonschema`
- **0.2.3**
  - Support generic sum types (aka sealed trait hierarchies)
  

core/src/main/scala/json/Schema.scala
package json

import com.github.andyglow.json.Value.ValueAdapter
import com.github.andyglow.json._
import com.github.andyglow.jsonschema.AsTree
import json.Schema.`object`.Field.RWMode
import json.schema.{validation => V}

import scala.collection.immutable.ListMap

sealed trait Schema[+T] {
  type Self <: Schema[T]

  private var _title: Option[String]                    = None
  private var _description: Option[String]              = None
  private var _discriminationKey: Option[String]        = None
  private var _validations: collection.Seq[V.Def[_, _]] = Seq.empty
  def jsonType: String
  def withValidation[TT >: T, B](v: V.Def[B, _], vs: V.Def[B, _]*)(implicit
    bound: V.Magnet[TT, B]
  ): Self = {
    val copy = this.duplicate()
    copy._validations = (v +: vs).foldLeft(_validations) { case (agg, v) =>
      bound.append(agg, v)
    }
    copy
  }
  def description: Option[String]       = _description
  def title: Option[String]             = _title
  def discriminationKey: Option[String] = _discriminationKey

  // NOTE: `.toSeq` is required for scala 2.13
  // otherwise we'll see
  // type mismatch;
  //  [error]  found   : Seq[json.ValidationDef[_, _]] (in scala.collection)
  //  [error]  required: Seq[json.ValidationDef[_, _]] (in scala.collection.immutable)
  def validations: Seq[V.Def[_, _]] = _validations.toSeq
  def toDebugString: String         = AsTree(this).rendered
  protected object ToString {
    def apply(fn: StringBuilder => Any): String = {
      val sb = new StringBuilder
      fn(sb)
      writeValidations(sb)
      writeExtra(sb)
      sb.toString
    }
    def writeExtra(sb: StringBuilder): Unit = {
      _description foreach { x => sb.append(" description=`").append(x).append('`') }
      _title foreach { x => sb.append(" title=`").append(x).append('`') }
      _discriminationKey foreach { x => sb.append(" discriminationKey=`").append(x).append('`') }
    }
    def writeValidations(sb: StringBuilder): Unit = {
      if (validations.nonEmpty) {
        sb.append(" {")
        var f = true
        validations foreach { v =>
          if (!f) sb.append(", ")
          sb.append(v.validation)
          sb.append(":=")
          sb.append(v.json)
          f = false
        }
        sb.append("}")
      }

      ()
    }
  }
  protected def mkCopy(): Self
  def duplicate(
    description: Option[String] = this._description,
    title: Option[String] = this._title,
    discriminationKey: Option[String] = this._discriminationKey
  ): Self = {

    val copy = mkCopy()
    copy._validations = this._validations
    copy._discriminationKey = discriminationKey
    copy._description = description
    copy._title = title

    copy
  }
  def withExtraFrom(x: Schema[_]): Self = {
    val copy = mkCopy()
    copy._validations = x._validations
    copy._description = x._description
    copy._title = x._title
    copy._discriminationKey = x._discriminationKey

    copy
  }
  def withValidationsAddedFrom(x: Schema[_]): Self = {
    val copy = mkCopy()
    copy._validations = copy._validations ++ x._validations

    copy
  }
  def canEqual(that: Any): Boolean = that.isInstanceOf[Schema[_]] // && getClass == that.getClass
  override def equals(obj: Any): Boolean = obj match {
    case s: Schema[_] =>
      s.canEqual(this) &&
      this.title == s.title &&
      this.description == s.description &&
      this.discriminationKey == s.discriminationKey &&
      // compare collections disregarding order
      this.validations.forall(s.validations.contains) &&
      s.validations.forall(this.validations.contains)

    case _ => false
  }
  def withDescription(x: String): Self                     = duplicate(description = Some(x))
  def withTitle(x: String): Self                           = duplicate(title = Some(x))
  def withDiscriminationKey(x: String): Self               = duplicate(discriminationKey = Some(x))
  def toDefinition[TT >: T](sig: String): Schema.`def`[TT] = Schema.`def`(sig, this)
  @deprecated("please use `toDefinition` instead", "1.0.0") def apply(refName: String): Schema[T] =
    toDefinition(refName)
}

object Schema {

  def apply[T: Schema]: Schema[T] = implicitly

  // +------------
  // | Boolean
  // +---------------
  //
  sealed class `boolean` extends Schema[Boolean] {
    type Self = `boolean`
    def mkCopy()                              = new `boolean`
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`boolean`]
    override def jsonType: String             = "boolean"
    override def toString: String             = ToString(_ append "boolean")
  }
  object `boolean` extends `boolean`

  // +------------
  // | Integer
  // +---------------
  //
  sealed class `integer` extends Schema[Int] {
    type Self = `integer`
    def mkCopy()                              = new `integer`()
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`integer`]
    override def jsonType: String             = "integer"
    override def toString: String             = ToString(_ append "integer")
  }
  object `integer` extends `integer`

  // +------------
  // | Number
  // +---------------
  //
  final class `number`[T: Numeric] extends Schema[T] {
    type Self = `number`[T]
    def mkCopy()                              = new `number`[T]
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`number`[_]]
    override def jsonType: String             = "number"
    override def toString: String             = ToString(_ append "number")
  }
  object `number` {
    def apply[T: Numeric]: `number`[T] = new `number`[T]
  }

  // +------------
  // | String
  // +---------------
  //
  sealed case class `string`[T](format: Option[`string`.Format]) extends Schema[T] {
    type Self = `string`[T]
    def mkCopy()                              = new `string`[T](format)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`string`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `string`(f) => format == f && super.equals(obj)
      case _           => false
    }
    override def jsonType: String = "string"
    override def toString: String = ToString { sb =>
      sb append "string"
      format foreach { format =>
        sb append "(format = "
        sb append format
        sb append ")"
      }
    }
  }
  object `string` extends `string`[String](None) {
    def apply[T]: `string`[T]                 = new `string`[T](None)
    def apply[T](format: Format): `string`[T] = new `string`[T](Some(format))
    trait Format extends Product
    object Format {
      case object `date`      extends Format
      case object `time`      extends Format
      case object `date-time` extends Format // Date representation, as defined by RFC 3339, section 5.6.
      case object `email`     extends Format // Internet email address, see RFC 5322, section 3.4.1.
      case object `hostname`  extends Format // Internet host name, see RFC 1034, section 3.1.
      case object `ipv4`      extends Format // Internet host name, see RFC 1034, section 3.1.
      case object `ipv6`      extends Format // IPv6 address, as defined in RFC 2373, section 2.2.
      case object `uri`       extends Format // A universal resource identifier (URI), according to RFC3986.

      // added in 2019-09
      case object `duration`     extends Format // The duration format is from the ISO 8601 ABNF as given in Appendix A of RFC 3339
      case object `idn-hostname` extends Format // Use RFC 1123 instead of RFC 1034; this allows for a leading digit,
      // `hostname` is also RFC 1123 since 2019-09
      case object `uuid` extends Format // A string instance is valid against this attribute if it is a valid string representation of a UUID, according to RFC4122
    }
  }

  // +------------
  // | Array
  // +---------------
  //
  final case class `array`[T, C[_]](componentType: Schema[T], unique: Boolean = false) extends Schema[C[T]] {
    type Self = `array`[T, C]
    def mkCopy() = new `array`[T, C](componentType, unique)
    override def canEqual(that: Any): Boolean = that match {
      case `array`(_, _) => true
      case _             => false
    }
    override def equals(obj: Any): Boolean = obj match {
      case `array`(c, u) => u == unique && componentType == c && super.equals(obj)
      case _             => false
    }
    override def jsonType: String = "array"
    override def toString: String = ToString { sb =>
      sb append "array(component ="
      sb append componentType
      sb append ", unique ="
      sb append unique
      sb append ")"
    }
  }

  // +------------
  // | Dictionary
  // +---------------
  //
  final case class `dictionary`[K, V, C[_, _]](valueType: Schema[V]) extends Schema[C[K, V]] {
    type Self = `dictionary`[K, V, C]
    override def jsonType = "object"
    def mkCopy()          = new `dictionary`[K, V, C](valueType)
    override def canEqual(that: Any): Boolean = that match {
      case `dictionary`(_) => true
      case _               => false
    }
    override def equals(obj: Any): Boolean = obj match {
      case `dictionary`(c) => valueType == c && super.equals(obj)
      case _               => false
    }
    override def toString: String = ToString { sb =>
      sb append "dictionary(value ="
      sb append valueType
      sb append ")"
    }
  }
  object `dictionary` {
    abstract class KeyPattern[T](val pattern: String)
    object KeyPattern {
      def mk[T](pattern: String): KeyPattern[T] = new KeyPattern[T](pattern) {}
      def forEnum[T](vals: Iterable[String]): KeyPattern[T] = {
        require(vals.nonEmpty)
        mk[T](vals.toList.distinct.sorted.mkString("^(?:", "|", ")$"))
      }
      implicit object StringKeyPattern extends KeyPattern[String]("^.*$")
      implicit object CharKeyPattern   extends KeyPattern[Char]("^.{1}$")
      implicit object ByteKeyPattern   extends KeyPattern[Byte]("^[0-9]+$")
      implicit object ShortKeyPattern  extends KeyPattern[Short]("^[0-9]+$")
      implicit object IntKeyPattern    extends KeyPattern[Int]("^[0-9]+$")
      implicit object LongKeyPattern   extends KeyPattern[Long]("^[0-9]+$")
    }
  }

  // +------------
  // | Object
  // +---------------
  //
  sealed case class `object`[T] private (fields: List[`object`.Field[_]]) extends Schema[T] {
    import `object`._
    type Self = `object`[T]
    def mkCopy()                              = copy()
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`object`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `object`(f) => fields.toSet == f.toSet && super.equals(obj)
      case _           => false
    }

    def dropField(pred: Field[_] => Boolean): `object`[T] =
      copy(fields = this.fields.filterNot(pred)).withExtraFrom(this)
    def withField(f: Field[_]): `object`[T]                                            = copy(fields = fields :+ f).withExtraFrom(this)
    def withField(name: String, tpe: Schema[_], required: Boolean = true): `object`[T] = withField(Field(name, tpe, required))
    def withFieldsUpdated(pf: PartialFunction[Field[_], Field[_]]): `object`[T] = copy(
      fields = fields collect {
        case f if pf isDefinedAt f => pf(f)
        case x                          => x
      }
    ).withExtraFrom(this)
    override def jsonType: String = "object"
    override def toString: String = ToString { sb =>
      sb append "object("
      fields foreach { field =>
        sb append field
        sb append ", "
      }
      sb.setLength(sb.length - 1) // drop last comma
      sb append ")"

      if (this.isInstanceOf[Free]) { sb append " additionalProperties=true" }
    }

    def free: `object`[T] with Free = {
      val self = this
      new `object`[T](fields) with Free {
        type Type = T
        def strict: `object`[T] = self
      }
    }
  }
  object `object` {
    sealed trait Free { this: `object`[_] =>
      type Type
      def strict: `object`[Type]
    }
    object Free {
      def apply[T](): `object`[T] with Free = {
        new `object`[T](List.empty) with Free {
          type Type = T
          override def strict: `object`[T] = new `object`[T](List.empty)
        }
      }
    }
    final case class Field[T](
      name: String,
      tpe: Schema[T],
      required: Boolean,
      default: Option[Value],
      description: Option[String],
      rwMode: Field.RWMode
    ) {
      def canEqual(that: Any): Boolean = that.isInstanceOf[Field[T]]
      override def equals(that: Any): Boolean = canEqual(that) && {
        val other = that.asInstanceOf[Field[T]]
        this.name == other.name &&
        this.required == other.required &&
        this.tpe == other.tpe &&
        this.default == other.default &&
        this.rwMode == other.rwMode
      }
      override def hashCode: Int = name.hashCode
      override def toString: String = {
        var extra = (required, default) match {
          case (true, None)     => " /R"
          case (false, None)    => ""
          case (true, Some(v))  => s" /R /$v"
          case (false, Some(v)) => s" /$v"
        }
        description foreach { x => extra = extra + s" description=`$x`" }
        s"$name: $tpe: $rwMode$extra"
      }
      def withDescription(x: Option[String]): Field[T] =
        new Field(name, tpe, required, default, x, rwMode)
      def withRWMode(x: RWMode): Field[T] = new Field(name, tpe, required, default, description, x)
      def setReadOnly: Field[T]           = withRWMode(RWMode.ReadOnly)
      def setWriteOnly: Field[T]          = withRWMode(RWMode.WriteOnly)
    }
    object Field {
      sealed trait RWMode
      object RWMode {
        case object ReadOnly  extends RWMode
        case object WriteOnly extends RWMode
        case object ReadWrite extends RWMode
      }

      def apply[T](name: String, tpe: Schema[T]): Field[T] =
        new Field(name, tpe, required = true, default = None, description = None, RWMode.ReadWrite)

      def apply[T](name: String, tpe: Schema[T], required: Boolean): Field[T] =
        new Field(name, tpe, required, default = None, description = None, RWMode.ReadWrite)

      def apply[T: ToValue](
        name: String,
        tpe: Schema[T],
        required: Boolean,
        default: T,
        rwMode: RWMode = RWMode.ReadWrite
      ): Field[T] =
        new Field(name, tpe, required, Some(ToValue(default)), description = None, rwMode = rwMode)

      def fromJson[T](
        name: String,
        tpe: Schema[T],
        required: Boolean,
        default: Option[Value],
        rwMode: RWMode = RWMode.ReadWrite
      ): Field[T] = new Field(name, tpe, required, default, description = None, rwMode = rwMode)
    }

    def apply[T](field: Field[_], xs: Field[_]*): `object`[T] = {
      fromList(field +: xs.toList)
    }

    def fromList[T](fields: List[Field[_]]): `object`[T] = {
      `object`(fields.distinct)
    }
  }

  // +------------
  // | Enum
  // +---------------
  //
  final case class `enum`[T](tpe: Schema[_], values: Set[Value]) extends Schema[T] {
    type Self = `enum`[T]
    def mkCopy()                              = new `enum`[T](tpe, values)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`enum`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `enum`(t, v) => t == tpe && values == v && super.equals(obj)
      case _            => false
    }
    override def jsonType: String = "enum"
    override def toString: String = ToString { sb =>
      sb append "enum["
      sb append tpe.toString
      sb append "]("
      values foreach { value =>
        sb append value
        sb append ","
      }
      sb.setLength(sb.length - 1) // drop last comma
      sb append ")"
    }
  }
  object `enum` {
    def of[T](tpe: Schema[_], x: Value, xs: Value*): `enum`[T] = new `enum`[T](tpe, (x +: xs).toSet)
    def of[T](x: T, xs: T*)(implicit va: ValueAdapter[T], vs: ValueSchema[T]): `enum`[vs.S] = {
      new `enum`[vs.S](vs.schema, (x +: xs).toSet.map { (x: T) => va.adapt(x) })
    }
  }

  // +------------
  // | OneOf
  // +---------------
  //
  final case class `oneof`[T](subTypes: Set[Schema[_]], discriminationField: Option[String] = None) extends Schema[T] {
    type Self = `oneof`[T]
    def mkCopy()                              = new `oneof`[T](subTypes, discriminationField)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`oneof`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `oneof`(s, d) => subTypes == s && discriminationField == d && super.equals(obj)
      case _             => false
    }
    override def jsonType: String = "oneof"
    override def toString: String = ToString { sb =>
      sb append "oneof("
      subTypes foreach { tpe =>
        sb append tpe
        sb append ","
      }
      sb.setLength(sb.length - 1) // drop last comma
      discriminationField foreach { f =>
        sb append "| discriminationField="
        sb append f
      }
      sb append ")"
    }
    def discriminatedBy(x: String): Self = new `oneof`[T](subTypes, Some(x))
  }
  object `oneof` {
    def of[T](x: Schema[_], xs: Schema[_]*): `oneof`[T] = new `oneof`[T]((x +: xs).toSet, None)
  }

  // +------------
  // | AllOf
  // +---------------
  //
  final case class `allof`[T](subTypes: Set[Schema[_]]) extends Schema[T] {
    type Self = `allof`[T]
    override def jsonType: String             = "allof"
    def mkCopy()                              = new `allof`[T](subTypes)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`allof`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `allof`(s) => subTypes == s && super.equals(obj)
      case _          => false
    }
    override def toString: String = ToString { sb =>
      sb append "allof("
      subTypes foreach { tpe =>
        sb append tpe
        sb append ","
      }
      sb.setLength(sb.length - 1) // drop last comma
      sb append ")"
    }
  }
  object `allof` {
    def of[T](x: Schema[_], xs: Schema[_]*): `allof`[T] = new `allof`[T]((x +: xs).toSet)
  }

  // +------------
  // | Not
  // +---------------
  //
  final case class `not`[T](tpe: Schema[T]) extends Schema[T] {
    type Self = `not`[T]
    override def jsonType: String             = "not"
    def mkCopy()                              = new `not`[T](tpe)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`not`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `not`(t) => tpe == t && super.equals(obj)
      case _        => false
    }
    override def toString: String = ToString { sb =>
      sb append "not("
      sb append tpe
      sb append ")"
    }
  }

  // +------------
  // | Def
  // +---------------
  //
  final case class `def`[T](sig: String, tpe: Schema[_]) extends Schema[T] {
    type Self = `def`[T]
    override def jsonType: String             = ??? // s"$$ref"
    def mkCopy()                              = new `def`[T](sig, tpe)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`def`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `def`(s, t) => sig == s && tpe == t && super.equals(obj)
      case _           => false
    }
    override def toString: String = ToString { sb =>
      sb append "def(signature = "
      sb append sig
      sb append ", schema = "
      sb append tpe
      sb append ")"
    }
    override def withValidation[TT >: T, B](v: V.Def[B, _], vs: V.Def[B, _]*)(implicit
      bound: V.Magnet[TT, B]
    ): `def`[T] = copy(tpe = tpe.asInstanceOf[Schema[TT]].withValidation(v, vs: _*)).withExtraFrom(this)
    override def toDefinition[TT >: T](sig: String): `def`[TT] = {
      def deepCopy(x: Schema[_]): Schema[_] = {
        val y = x match {
          case `object`(fields)     => `object`(fields.map { f => f.copy(tpe = deepCopy(f.tpe)) })
          case `array`(y, u)        => `array`(deepCopy(y), u)
          case `dictionary`(y)      => `dictionary`(deepCopy(y))
          case `oneof`(ys, df)      => `oneof`(ys map deepCopy, df)
          case `allof`(ys)          => `allof`(ys map deepCopy)
          case `not`(y)             => `not`(deepCopy(y))
          case `def`(s, y)          => `def`(sig, deepCopy(y))
          case `ref`(s) if s == sig => `ref`(sig)
          case y                    => y
        }
        y withExtraFrom x
      }

      copy(sig = sig, tpe = deepCopy(tpe)).withExtraFrom(this)
    }
  }

  object `def` {
    def adapt[T](tpe: Schema[_], sig: => String): `def`[T] = {
      tpe match {
        case `def`(originalSig, innerTpe) => `def`(originalSig, innerTpe)
        case `value-class`(innerTpe)      => `def`(sig, innerTpe)
        case _                            => `def`(sig, tpe)
      }
    }
  }

  // +------------
  // | Value-Class
  // +---------------
  // Pseudo member, doesn't have it's own representation in resulted schema
  //
  final case class `value-class`[O, I](tpe: Schema[I]) extends Schema[O] {
    type Self = `value-class`[O, I]
    override def jsonType: String             = tpe.jsonType
    def mkCopy()                              = new `value-class`[O, I](tpe)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`value-class`[_, _]]
    override def equals(obj: Any): Boolean = obj match {
      case `value-class`(t) => tpe == t && super.equals(obj)
      case _                => false
    }
    override def toString: String = ToString { sb =>
      sb append "value-class("
      sb append tpe
      sb append ")"
    }
    override def toDefinition[TT >: O](sig: String): `def`[TT] =
      `def`[TT](sig, tpe.withValidationsAddedFrom(this))
  }

  // +------------
  // | LazyRef
  // +---------------
  // Pseudo member, doesn't have it's own representation in resulted schema
  //
  final case class `ref`[T](sig: String) extends Schema[T] {
    override type Self = `ref`[T]
    override protected def mkCopy(): `ref`[T] = `ref`(sig)
    override def jsonType: String =
      ??? // should never call this. instead calling code should interpret it as `ref`
    override def toString: String = ToString { sb =>
      sb append "ref("
      sb append sig
      sb append ")"
    }
  }

  // +------------
  // | Const
  // +---------------
  // It is used in conjunction with OneOf as Alternative to Enum
  // INTERNAL API
  final case class `const`[T](value: Value) extends Schema[T] {
    override def jsonType: String = "const"
    override type Self = `const`[T]
    override protected def mkCopy(): `const`[T] = `const`(value)
    override def toString: String = ToString { sb =>
      sb append "const("
      sb append JsonFormatter.format(value)
      sb append ": "
      sb append value.tpe
      sb append ")"
    }
    override def canEqual(that: Any): Boolean = that.isInstanceOf[`const`[_]]
    override def equals(obj: Any): Boolean = obj match {
      case `const`(v) => value == v && super.equals(obj)
      case _          => false
    }
  }
}
